Add support for Remix v2 (#215)
sergiodxa/remix-utils: README.md

This part of Remix Utils needs React and server-side code.

First create a new CSRF instance.

```ts
// app/utils/csrf.server.ts
import { CSRF } from "remix-utils";
import { createCookie } from "@remix-run/node"; // or /cloudflare

export const cookie = createCookie("csrf", {
  path: "/",
  httpOnly: true,
  secure: process.env.NODE_ENV === "production",
  sameSite: "lax",
  secrets: ["s3cr3t"],
});

export const csrf = new CSRF({
  cookie,
  // what key in FormData objects will be used for the token, defaults to `csrf`
  formDataKey: "csrf",
  // an optional secret used to sign the token, recommended for extra safety
  secret: "s3cr3t",
});
```

Then you can use `csrf` to generate a new token.

```ts
import { csrf } from "~/utils/csrf.server";

export async function loader({ request }: LoaderArgs) {
  let token = csrf.generate();
}
```

You can customize the token size by passing the byte size, the default one is 32 bytes which will give you a string with a length of 43 after encoding.

```ts
let token = csrf.generate(64); // customize token length
```

> You will need to save this token in a cookie and also return it from the loader, for convenience, you can use the `CSRF#commitToken` helper.

Reviewer: ```suggestion
You will need to save this token in a cookie and also return it from the loader. For convenience, you can use the `CSRF#commitToken` helper.
```

import { useAuthenticityToken } from "remix-utils";

export function useMarkAsRead() {
  let fetcher = useFetcher();
  let csrf = useAuthenticityToken();
  return function submit(data) {
    fetcher.submit(
      { csrf, ...data },
      { action: "/api/mark-as-read", method: "post" }
    );
  };
}
```

Finally, you need to validate the authenticity token in the action that received the request.

```ts
import { CSRFError, redirectBack } from "remix-utils";
import { csrf } from "~/utils/csrf.server";

export async function action({ request }: ActionArgs) {
  try {
    await csrf.validate(request);
  } catch (error) {
    if (error instanceof CSRFError) {
      // handle CSRF errors
    }
    // handle other possible errors
  }

  // here you know the request is valid
  return redirectBack(request, { fallback: "/fallback" });
}
```

If you need to parse the body as FormData yourself (e.g. to support file uploads) you can also call `CSRF#validate` with the FormData and Headers objects.

```ts
let formData = await parseMultiPartFormData(request);
try {
  await csrf.validate(formData, request.headers);
} catch (error) {
  // handle errors
}
```

> > **Warning**: If you call `CSRF#validate` with the request instance but you read the body before it will throw an error.

Reviewer: ```suggestion
> **Warning**: If you call `CSRF#validate` with the request instance, but you already read its body, it will throw an error.
```

> In case the CSRF validation fails, it will throw a `CSRFError` which can be used to correctly identify them against other possible errors that may throw.

Reviewer: ```suggestion
In case the CSRF validation fails, it will throw a `CSRFError` which can be used to correctly identify it against other possible errors that may get thrown.
```

> Tip: You could use it together with useShouldHydrate to disable Remix scripts in certain routes but still load scripts for analytics or small features that need JS but don't need the full app JS to be enabled.

### useGlobalNavigationState

> This hook lets you know if the value of `transition.state`, every `fetcher.state` in the app, and `revalidator.state`.

Reviewer: ```suggestion
This hook allows you to read the value of `transition.state`, every `fetcher.state` in the app, and `revalidator.state`.
```

Reviewer: not sure about this one, but the original sentence definitely needs to be rewritten

}
```

The return value of `useGlobalNavigationState` can be `"idle"`, `"loading"` or `"submitting"`

> **Note** This is used by the hooks below to determine if the app is loading, submitting or both (pending).

### useGlobalPendingState

> This hook lets you know if the global navigation, if one of any active fetchers is either loading or submitting, or the revalidator is running.

Reviewer: ```suggestion
This hook lets you know if the global navigation, if one of any active fetchers is either loading or submitting, or if the revalidator is running.
```

### useGlobalLoadingState

> This hook lets you know if the global transition, if one of any active fetchers is loading, or the revalidator is running

Reviewer: ```suggestion
This hook lets you know if the global transition, if one of any active fetchers is loading, or if the revalidator is running
```

For Server-Sent Events to work, your server must support HTTP streaming. If you don't get SSE to work check if your deployment platform has support for it.

> Because SSE count towards the limit of HTTP connections per domain, the useEventSource hook keeps a global map of connections based on the provided URL and options, as long as they are the same the hook will open a single SSE connection and share it between instances of the hook.

Reviewer: ```suggestion
Because SSE count towards the limit of HTTP connections per domain, the useEventSource hook keeps a global map of connections based on the provided URL and options, as long as they are the same, the hook will open a single SSE connection and share it between instances of the hook.
```

> Once there are not more instances of the hook re-using a connection, the connection will be closed and removed from the map.

Reviewer: ```suggestion
Once there are no more instances of the hook re-using a connection, it will be closed and removed from the map.
```

For Server-Sent Events to work, your server must support HTTP streaming. If you don't get SSE to work check if your deployment platform has support for it.

> Because SSE count towards the limit of HTTP connections per domain, the useEventSource hook keeps a global map of connections based on the provided URL and options, as long as they are the same, the hook will open a single SSE connection and share it between instances of the hook.

Reviewer: ```suggestion
Because SSE count towards the limit of HTTP connections per domain, the `useEventSource` hook keeps a global map of connections based on the provided URL and options. As long as they are the same, the hook will open a single SSE connection and share it between instances of the hook.
```

Once there are no more instances of the hook re-using a connection, it will be closed and removed from the map.

You can use the `<EventSourceProvider />` component to control the map.

```tsx
let map: EventSourceMap = new Map();
return (
  <EventSourceProvider value={map}>
    <YourAppOrPartOfIt />
  </EventSourceProvider>
);
```

> This way, you could overwrite the map with a new one on a part of the app. Note that this provider is optional and a default map will be used if you don't provide one.

Reviewer: ```suggestion
This way, you can overwrite the map with a new one for a specific part of your app. Note that this provider is optional and a default map will be used if you don't provide one.
```
